Source/Urho3D/Physics/TriggerAnimator.cpp

```diff
@@ -0,0 +1,208 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+#pragma once
+
+#include "Urho3D/Core/Context.h"
+#include "Urho3D/Graphics/Animation.h"
+#include "Urho3D/Physics/PhysicsEvents.h"
+#include "Urho3D/Physics/TriggerAnimator.h"
+
+#include "Urho3D/Graphics/AnimationController.h"
+#include "Urho3D/Resource/ResourceCache.h"
+
+namespace Urho3D
+{
+
+TriggerAnimator::TriggerAnimator(Context* context)
+    : BaseClassName(context)
+{
+}
+
+TriggerAnimator::~TriggerAnimator()
+{
+}
+
+void TriggerAnimator::RegisterObject(Context* context)
+{
+    context->AddFactoryReflection<TriggerAnimator>(Category_Physics);
+
+    URHO3D_MIXED_ACCESSOR_ATTRIBUTE("Enter Animation", GetEnterAnimationAttr, SetEnterAnimationAttr, ResourceRef,
+        ResourceRef(Animation::GetTypeStatic()), AM_DEFAULT);
+    URHO3D_MIXED_ACCESSOR_ATTRIBUTE("Exit Animation", GetExitAnimationAttr, SetExitAnimationAttr, ResourceRef,
+        ResourceRef(Animation::GetTypeStatic()), AM_DEFAULT);
+    URHO3D_ATTRIBUTE("Is Entered", bool, isEntered_, false, AM_DEFAULT);
+    
+}
+
+void TriggerAnimator::SetEnterAnimationAttr(const ResourceRef& value)
+{
+    auto* cache = GetSubsystem<ResourceCache>();
+    enterAnimation_ = cache->GetResource<Animation>(value.name_);
+}
+
+ResourceRef TriggerAnimator::GetEnterAnimationAttr() const
+{
+    return GetResourceRef(enterAnimation_, Animation::GetTypeStatic());
+}
+
+void TriggerAnimator::SetExitAnimationAttr(const ResourceRef& value)
+{
+    auto* cache = GetSubsystem<ResourceCache>();
+    exitAnimation_ = cache->GetResource<Animation>(value.name_);
+}
+
+ResourceRef TriggerAnimator::GetExitAnimationAttr() const
+{
+    return GetResourceRef(exitAnimation_, Animation::GetTypeStatic());
+}
+
+void TriggerAnimator::OnEnter()
+{
+}
+
+void TriggerAnimator::OnExit()
+{
+}
+
+void TriggerAnimator::OnNodeSet(Node* previousNode, Node* currentNode)
+{
+    Component::OnNodeSet(previousNode, currentNode);
+    UpdateSubscriptions();
+}
+
+void TriggerAnimator::OnSetEnabled()
+{
+    Component::OnSetEnabled();
+    UpdateSubscriptions();
+}
+
+void TriggerAnimator::RegisterEnter(Node* node)
+{
+    if (!node)
+        return;
+    if (activeCollisions_.empty())
+    {
+        if (!isEntered_)
+        {
+            isEntered_ = true;
+            StartAnimation(enterAnimation_);
+            OnEnter();
+        }
+    }
+    activeCollisions_.insert(node);
+}
+
+void TriggerAnimator::RegisterExit(Node* node)
+{
+    if (!node)
+        return;
+    activeCollisions_.erase(node);
+    if (activeCollisions_.empty())
+    {
+        if (isEntered_)
+        {
+            isEntered_ = false;
+            StartAnimation(exitAnimation_);
+            OnExit();
+        }
+    }
+}
+
+void TriggerAnimator::HandlePhysicsCollisionStart(StringHash eventType, VariantMap& eventData)
```

Comment:
I don't like M*N complexity. All triggers process all collisions and then ignore most of them.
They should subscribe only to relevant Node instead.
Other than that, it looks okay.